[DOCS] Add guidance for mapping unstructured content

jrodewig opened the pull request:
Fixes #69046

### Preview
https://elasticsearch_69079.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/keyword.html#wildcard-field-type

jrodewig committed:
[DOCS] Add guidance for wildcard field performance

elasticmachine commented:
Pinging @elastic/es-search (Team:Search)

elasticmachine commented:
Pinging @elastic/es-docs (Team:Docs)

jrodewig committed:
Note storage costs for shared prefixes

jrodewig commented:
Hi @markharwood! Let me know what you think of these changes. I'm happy to iterate if needed. Thanks!

markharwood commented on docs/reference/mapping/types/keyword.asciidoc:
```diff
@@ -11,8 +11,10 @@ The keyword family includes the following field types:
-for grep-like <<query-dsl-wildcard-query,wildcard queries>>.
+* <<wildcard-field-type,`wildcard`>>, which optimizes high-cardinality keyword
+fields, such as log messages, for grep-like
+<<query-dsl-wildcard-query,`wildcard`>> and <<query-dsl-regexp-query,`regexp`>>
```
>optimizes high-cardinality keyword fields, such as log messages,

My comment below is long and the TL/DR is that different people have different notions of what `keyword` fields are used for so it's perhaps unwise to make comparisons with it. We need to target advice to these different camps.

Part of the problem is we badged wildcard field as a better keyword field. I think in practice it's a better `text` field for non-human-authored large strings like log messages.

* In ECS they actually map log messages as a text field (which can be tough to query but fast).
* Some [3rd party security systems](https://socprime.com/blog/elastic-for-security-analysts-part-1-searching-strings/) mapped log messages as keyword fields because they were easier to query (but slow and missing large values).

So the assumption that log messages == keyword fields is not one that is universally shared.

Keyword fields are only like wildcard fields in that they don't chop strings into words - it's conceptually a single value. That is more about how they are implemented rather than how they are used. Classic keyword fields I tend to think of as short, structured values like country code or customer IDs. They are used to sort and aggregate on on and nearly always queried using exact **full**-values.

The [3rd party security systems](https://socprime.com/blog/elastic-for-security-analysts-part-1-searching-strings/) (ab)use of keyword fields for log messages was to overcome a **partial**-match search problem - not for their strengths in structured aggs, sorting or fast full-value-match search capabilities.

Why are wildcard fields more like a `text` field?  It's typically for large and unstructured content where you only want to search for bits of it (no aggs or sorting required). However, it's a different form of partial search to classic text-matching because you look for character sequences, not word sequences (there's no pre-determined opinion on what constitutes a "word" in the string because of Analyzers). 

In my view our previous advice that `wildcard == keyword` field was misleading and consequently the ECS team replaced many of their classic short, structured keyword fields with wildcard when it was unsuitable and this change had to be backed out. The disk costs and search speeds were worse than the classic keyword field. Ironically, the one field in ECS that might have benefitted from the change was the log `message` field and that stayed mapped as a `text` field. Changing this to the wildcard field however would be a big UX change - people would have to learn to search for character sequence`*foo*` rather than just the plain word `foo`. The problem with `foo` searches in `text` log messages was always that the tokenizer might consider `org.elastic.foo.bar` as one word and not match. _Word-based searches only work well with human language, and not so well for computer-speak._

Targeted advice:
For people considering switching from `text` to `wildcard` (like ECS ) we should document the necessary  change in UX and how word-based text matching like `quick AND brown AND fox` needs to change to equivalent queries for character sequences like `[Qq]uick\wbrown\wfox` (it doesn't help that KQL doesn't support regex currently).

For people like the 3rd party security folks who already use regexes and wildcard queries they have essentially been deploying a hack putting log messages into keyword fields because of their frustrations with text Analyzers. They need a different message (with wildcard field your query UX stays the same, but search speeds can be improved if log messages are many and varied plus we can deal with longer messages).

jrodewig replied:
Thanks for the explanation @markharwood. I'll work on an update to address this.

markharwood reviewed (commented):
Many thanks for this PR. I added some rambling comments. Unfortunately this whole wildcard topic still feels like one big "it depends" essay and I struggle to distil it down to some short guidance.
I'm hoping you might be able to shape my concerns into something much more digestable.

jrodewig committed:
Fix Boolean criteria

jrodewig commented on docs/reference/mapping/types/wildcard.asciidoc:
```diff
@@ -4,16 +4,40 @@
+* The field contains values larger than 32KB. +
+OR +
+The field contains more than a million unique values.
+
```
Thanks for the clarification. I've updated this with da634aa4867.

jrodewig committed:
Add sidebar for unstructured machine-gen content

jrodewig committed:
Reword

jrodewig committed:
Clarifications

jrodewig committed:
Update bullet

jrodewig committed:
Rework

jrodewig commented:
Thanks again for the thorough explanation @markharwood.

I reworked this section to better cover the concerns you raised. As part of that, I added a new sidebar for mapping unstructured content. The idea is to give the user clear, concise, and actionable advice on which field type to use depending on their use case. The sidebar also includes a section on switching from a `text` field to a `wildcard` or `keyword` field.

Let me know if you have any other thoughts. Thanks!

markharwood commented:
>I reworked this section to better cover the concerns you raised. As part of that, I added a new sidebar for mapping unstructured content.

That looks great, James. Many thanks for the work on this
